src/dataflow/src/source/file.rs

```diff
@@ -311,28 +311,39 @@ fn open_file_stream(
                     // what we do too.
                     let mut buf = [0; 1024];
                     loop {
-                        if let Err(err) = inotify.read_events_blocking(&mut buf) {
-                            if notice_tx
-                                .send(Err(format!(
+                        match inotify.read_events_blocking(&mut buf) {
+                            Err(err) => {
+                                if notice_tx
+                                    .send(Err(format!(
                                     "file source: failed to get events for file: {:#} (path: {})",
                                     err,
                                     _path.display()
                                 )))
-                                .is_err()
-                            {
-                                // If the notice_tx returns an error, it's because
-                                // the source has been dropped. Just exit the
-                                // thread.
+                                    .is_err()
+                                {
+                                    // If the notice_tx returns an error, it's because
+                                    // the source has been dropped. Just exit the
+                                    // thread.
+                                    return;
+                                }
+                                // We have no method for recovering from this error
+                                // Close this thread and log an error message (which duplicates the err above)
+                                error!(
+                                    "file source: closing stream due to read errors (path: {})",
+                                    _path.display()
+                                );
                                 return;
                             }
-                            // We have no method for recovering from this error
-                            // Close this thread and log an error message (which duplicates the err above)
-                            error!(
-                                "file source: closing stream due to read errors (path: {})",
-                                _path.display()
-                            );
-                            return;
-                        };
+                            Ok(mut events) => {
```

Comment:
I think it's because `Events` type implements `Iterator` but its `Item` is of type `Events` rather than `&Events`, which doesn't implement `Copy`, so the iterator is destructive.